sys/exec.c

```diff
@@ -105,12 +105,21 @@ int do_exec(const exec_args_t *args) {
    * We can not destroy the current vm map, because exec can still fail,
    * and in that case we must be able to return to the original address space.
    */
-  vm_map_t *vmap = vm_map_new();
-  vm_map_t *old_vmap = p->p_uspace;
+  struct {
+    vm_map_t *uspace;
+    vm_map_entry_t *sbrk;
+    vm_addr_t sbrk_end;
+  } old = {p->p_uspace, p->p_sbrk, p->p_sbrk_end};
 
   /* We are the only live thread in this process. We can safely give it a new
    * uspace. */
+  vm_map_t *vmap = vm_map_new();
+
   p->p_uspace = vmap;
+  p->p_sbrk = NULL;
+  p->p_sbrk_end = 0;
+  sbrk_attach(p); /* Attach fresh brk segment. */
```

Comment:
Please note that this PR is named in somewhat misleading way.

The change is mostly concerned about attaching fresh brk segment to new process image.